### ISSUE TITLE
torch.distributed and horovod gave different results. The results using horovod are more unstable.

### ISSUE THREAD

**Issue by Fangyh09** [horovod/horovod#1567]
Hello, I tried horovod and torch.distributed.  The difference is a little big. 
The results using horovod is more unstable.  I tried three times for each.
Currently, I am trying to train each 10 times.

What are the possible reasons? 
horovod: 4 gpus x 8 batchsize
torch.distributed: 4 gpus x 8 batchsize

**Comment by abditag2** [comment]
Can you further clarify your question? Different results in model convergence or number of samples per second of training?

**Comment by abditag2** [comment]
And what is your model?

**Comment by tgaddair** [comment]
Hey @Fangyh09, do you have a reproducible example we can take a look at?  We'd be happy to take a look.

**Comment by Fangyh09** [comment]
Hello, currently more experiments are carried out.
I will summary my results later.  Currently, I think it is related to  https://github.com/horovod/horovod/issues/1548.

Thanks for your quick reply.

**Comment by tgaddair** [comment]
One thing to consider trying if you're seeing unstable gradients is to broadcast model state periodically, like once per epoch or more.  That can help if you're seeing significant drift between model replicas due to nondeterminism.

**Comment by Richie-yan** [comment]
@Fangyh09 This is because tensor_fusion of Horovod is dynamic. When using nccl to do allreduce (nccl divides the data into blocks), the order of adding floating point numbers may be inconsistent, which eventually leads to unstable gradients.
@tgaddair Does the horovod community consider changing tensor_fusion to static, like pytorch's DDP, allocate buffers according to model parameters before training?

**Comment by Fangyh09** [comment]
@Richie-yan 
Does these two things are equal?
1. horovod
2. torch.distribute, having set torch.backends.cudnn.benchmark = True, torch.backends.cudnn.deterministic = False

**Comment by Richie-yan** [comment]
@Fangyh09 
According to my experience, your second two settings are invalid for the instability of the gradient. It is not introduced by torch. As I said before, it is caused by the dynamic tensor_fusion of horovod (the most fundamental reason Is caused by the inconsistent sum order of floating point numbers)

**Comment by tgaddair** [comment]
@Richie-yan, @romerojosh is working on a PR that will allow you to control the level of determinism in the summation order (#1130).

**Comment by Fangyh09** [comment]
@Richie-yan 
In the setting two, horovod is not used.

**Comment by Fangyh09** [comment]
Besides the instability problems, what are the other possible reasons?

**Comment by Fangyh09** [comment]
I found another reason, it may be related to sync batchnorm.
